### PR TITLE
[ci][test-states/03] create github issues for failing tests

### PR DESCRIPTION
**Pull request by can-anyscale**
## Why are these changes needed?
Context: This is a series of PR to compute test state (JAILED|PASSING|FAILING) on-the-fly, persist them into DB and create issues and bisect automatically, etc.

- Add hooks to trigger actions when tests move to a different state. 
- Add an action to create github issue when test moves from PASSING to FAILING state

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [X] Release tests
   - intentionally cause remove_before_commit test to fail
   - observe that github issue is created correctly: https://github.com/ray-project/ray/issues/35896

### CODE REVIEW

File: release/ray_release/test_automation/state_machine.py [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
     Test,
     TestState,
 )
 
+RAY_REPO = "ray-project/ray"
+AWS_SECRET_GITHUB = "ray_ci_github_token"
+
 
 class TestStateMachine:
+    ray_repo = None
```
Comment (aslonnie) [inline release/ray_release/test_automation/state_machine.py:14]: this is a class attribute, which is shared for all the instances of this class.

which means that when you have multiple of `TestStateMachine`, line 33's check and lazy set can be a racing condition where the secret gets read twice.

so it is a pattern that is a bit confusing.

Comment (aslonnie) [reply release/ray_release/test_automation/state_machine.py:14]: if you want to stub/fake `ray_repo`, you can send in an optional arg on construction, or maybe just alternate `self.ray_repo` directly after init.

File: release/ray_release/test_automation/state_machine.py [hunk 2]
```diff
@@ -22,13 +30,46 @@ def __init__(self, test: Test) -> None:
...
+        to_state = self.state_machine[from_state]
+        self.test.set_state(to_state)
+        self._move_hook(from_state, to_state)
+
+    def _move_hook(self, from_state: TestState, to_state: TestState) -> None:
+        """
+        Action performed when test transitions to a different state. This is where we do
+        things like creating and closing github issues, trigger bisects, etc.
+        """
+        if from_state == TestState.PASSING and to_state == TestState.FAILING:
```
[the remark this answers is not in the document]
Comment (can-anyscale) [reply release/ray_release/test_automation/state_machine.py:53]: i like that

File: release/release_tests.yaml [hunk 3]
```diff
@@ -1714,6 +1714,21 @@
 #######################
 # Tune cloud  tests
 #######################
+- name: remove_before_commit
```
Comment (can-anyscale) [inline release/release_tests.yaml:1717]: just here for temporary testing, will remove before commit

**Review by aslonnie — APPROVED** [review]

**Comment by ollie-iterators** [comment]
This seems like an important PR to get merged.

File: release/release_tests.yaml (hunk 3 above)
Comment (can-anyscale) [reply release/release_tests.yaml:1717]: actually will keep it in prod for testing and will remove it later

**Review by richardliaw — APPROVED** [review]

**Pull request merged**